File: voice-memo-watch/watchos/VoiceMemoBot/APIClient.swift
import Foundation

enum APIError: Error {
    case badResponse
}

/// URLSession client for the VoiceMemoBot backend. The backend serializes
/// camelCase JSON, so the default JSONDecoder maps straight onto Models.swift.
struct APIClient {
    var baseURL = AppConfig.apiBaseURL

    func fetchStyles() async throws -> [Style] {
        try await get([Style].self, path: "/styles")
    }

    func fetchInstruments() async throws -> [Instrument] {
        try await get([Instrument].self, path: "/instruments")
    }

    func fetchPromptSuggestions() async throws -> [PromptSuggestion] {
        try await get([PromptSuggestion].self, path: "/prompts/suggestions")
    }

    func uploadMemo(fileURL: URL) async throws -> Memo {
        let boundary = "vmb-\(UUID().uuidString)"
        var request = URLRequest(url: baseURL.appendingPathComponent("/memos"))
        request.httpMethod = "POST"
        request.setValue(
            "multipart/form-data; boundary=\(boundary)",
            forHTTPHeaderField: "Content-Type"
        )

        var body = Data()
        body.append("--\(boundary)\r\n")
        body.append(
            "Content-Disposition: form-data; name=\"file\"; filename=\"memo.m4a\"\r\n"
        )
        body.append("Content-Type: audio/mp4\r\n\r\n")
        body.append(try Data(contentsOf: fileURL))
        body.append("\r\n--\(boundary)--\r\n")

        let (data, response) = try await URLSession.shared.upload(
            for: request, from: body
        )
        try Self.ensureSuccess(response)
        return try JSONDecoder().decode(Memo.self, from: data)
    }

    func renderMemo(
        memoId: String,
        style: String,
        tweaks: Tweaks = Tweaks(),
        instruments: [String] = [],
        prompt: String = ""
    ) async throws -> Render {
        try await post(
            Render.self,
            path: "/memos/\(memoId)/renders",
            body: RenderRequest(
                style: style, tweaks: tweaks, instruments: instruments, prompt: prompt
            )
        )
    }

    func createPost(
        renderId: String, author: String, caption: String
    ) async throws -> Post {
        try await post(
            Post.self,
            path: "/posts",
            body: PostRequest(renderId: renderId, author: author, caption: caption)
        )
    }

    func fetchFeed() async throws -> [Post] {
        try await get([Post].self, path: "/posts")
    }

    func likePost(postId: String) async throws -> Post {
        try await post(Post.self, path: "/posts/\(postId)/like", body: nil as Int?)
    }

    private func get<T: Decodable>(_ type: T.Type, path: String) async throws -> T {
        let (data, response) = try await URLSession.shared.data(
            from: baseURL.appendingPathComponent(path)
        )
        try Self.ensureSuccess(response)
        return try JSONDecoder().decode(type, from: data)
    }

    private func post<T: Decodable, Body: Encodable>(
        _ type: T.Type, path: String, body: Body?
    ) async throws -> T {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = "POST"
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONEncoder().encode(body)
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        try Self.ensureSuccess(response)
        return try JSONDecoder().decode(type, from: data)
    }

    private static func ensureSuccess(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse,
              (200..<300).contains(http.statusCode)
        else { throw APIError.badResponse }
    }
}

private extension Data {
    mutating func append(_ string: String) {
        append(Data(string.utf8))
    }
}

File: voice-memo-watch/wearos/app/src/main/java/com/voicememobot/wear/ApiClient.kt
package com.voicememobot.wear

import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody

class ApiException(message: String) : Exception(message)

/**
 * OkHttp client for the VoiceMemoBot backend — the Kotlin twin of the
 * watchOS APIClient.swift, speaking the same camelCase JSON contract.
 */
class ApiClient(private val baseUrl: String = AppConfig.API_BASE_URL) {
    private val http = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }
    private val jsonMedia = "application/json".toMediaType()

    suspend fun fetchStyles(): List<Style> = get("/styles")

    suspend fun fetchInstruments(): List<Instrument> = get("/instruments")

    suspend fun fetchFeed(): List<Post> = get("/posts")

    suspend fun fetchPromptSuggestions(): List<PromptSuggestion> =
        get("/prompts/suggestions")

    suspend fun uploadMemo(file: File): Memo = withContext(Dispatchers.IO) {
        val body = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart(
                "file", "memo.m4a",
                file.asRequestBody("audio/mp4".toMediaType())
            )
            .build()
        execute(Request.Builder().url("$baseUrl/memos").post(body).build())
    }

    suspend fun renderMemo(
        memoId: String,
        style: String,
        tweaks: Tweaks = Tweaks(),
        instruments: List<String> = emptyList(),
        prompt: String = "",
    ): Render = postJson(
        "/memos/$memoId/renders",
        json.encodeToString(
            RenderRequest.serializer(),
            RenderRequest(style, tweaks, instruments, prompt)
        )
    )

    suspend fun createPost(renderId: String, author: String, caption: String): Post =
        postJson(
            "/posts",
            json.encodeToString(
                PostRequest.serializer(),
                PostRequest(renderId, author, caption)
            )
        )

    suspend fun likePost(postId: String): Post = withContext(Dispatchers.IO) {
        execute(
            Request.Builder()
                .url("$baseUrl/posts/$postId/like")
                .post(ByteArray(0).toRequestBody(null))
                .build()
        )
    }

    private suspend inline fun <reified T> get(path: String): T =
        withContext(Dispatchers.IO) {
            execute(Request.Builder().url("$baseUrl$path").get().build())
        }

    private suspend inline fun <reified T> postJson(path: String, payload: String): T =
        withContext(Dispatchers.IO) {
            val body: RequestBody = payload.toRequestBody(jsonMedia)
            execute(Request.Builder().url("$baseUrl$path").post(body).build())
        }

    private inline fun <reified T> execute(request: Request): T {
        http.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException("HTTP ${response.code}: $text")
            }
            return json.decodeFromString(text)
        }
    }
}
